python/ray/train/_internal/checkpoint_manager.py

```diff
@@ -0,0 +1,184 @@
+import logging
+import numbers
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from ray._private.dict import flatten_dict
+from ray.air.config import MAX
+from ray.air._internal.util import is_nan
+from ray.train import CheckpointConfig
+from ray.train._internal.storage import _delete_fs_path
+from ray.train.checkpoint import Checkpoint
+
+
+logger = logging.getLogger(__name__)
+
+
+class _TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager."""
+
+    def __init__(self, checkpoint: Checkpoint, metrics: Dict[str, Any]):
+        self.checkpoint = checkpoint
+        self.metrics = metrics
+
+
+def _insert_into_sorted_list(list: List[Any], item: Any, key: Callable[[Any], Any]):
+    """Insert an item into a sorted list with a custom key function.
+
+    Examples:
+
+        >>> list = []
+        >>> _insert_into_sorted_list(list, {"a": 1, "b": 0}, lambda x: x["a"])
+        >>> list
+        [{'a': 1, 'b': 0}]
+        >>> _insert_into_sorted_list(list, {"a": 3, "b": 1}, lambda x: x["a"])
+        >>> list
+        [{'a': 1, 'b': 0}, {'a': 3, 'b': 1}]
+        >>> _insert_into_sorted_list(list, {"a": 4, "b": 2}, lambda x: x["a"])
+        >>> list
+        [{'a': 1, 'b': 0}, {'a': 3, 'b': 1}, {'a': 4, 'b': 2}]
+        >>> _insert_into_sorted_list(list, {"a": 1, "b": 3}, lambda x: x["a"])
+        >>> list
+        [{'a': 1, 'b': 0}, {'a': 1, 'b': 3}, {'a': 3, 'b': 1}, {'a': 4, 'b': 2}]
+    """
+    i = 0
+    while i < len(list):
```

Comment:
Technically you could also use bisect, though I think this is also fine.

Reply:
Ah yeah I tried using bisect, but they don't support key based "insort" until python 3.10.